keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -0,0 +1,1018 @@
+2. Grant the special user permission to ONLY operate on the ephemeral containers subresource (it is
+   critical that they cannot create or update pods directly).
+3. Grant (real) users that should have privileged debug capability the ability to impersonate the
+   exempt user.
+
+We could consider ways to streamline the user experience of this, for instance adding a special RBAC
+binding that exempts users when operating on the ephemeral containers subresource (e.g. an
+`escalate-privilege` verb on the ephemeral containers subresource).
+
+<<[/UNRESOLVED]>>
+
+#### Other Pod Subresources
+
+Aside from ephemeral containers, the policy is not checked for any other Pod subresources (status,
+bind, logs, exec, attach, port-forward).
+
+Although annotations can be updated through the status subresource, the apparmor annotations are
+immutable and the seccomp annotations are deprecated and slated for removal in v1.23.
+
+### Pod Security Standards
+
+Policy level definitions are hardcoded and unconfigurable out of the box. However, the [Pod Security
+Standards] leave open ended guidance on a few items, so we must make a static decision on how to
+handle these elements:
+
+**HostPorts** - (baseline) HostPorts will be forbidden. This is a more niche feature, and violates
+the container-host boundary.
+
+**AppArmor** - (baseline) Allow anything except `unconfined`. Custom AppArmor profiles must be
+installed by the cluster admin, and AppArmor fails closed when a profile is not present.
+
+<<[UNRESOLVED]>>
+**BLOCKING**
+
+**SELinux** - (baseline)
+- Unset `SELinuxOptions` is allowed.
+- If options are set, they must match the following:
+  - `type` is one of:
+    - `container_init_t`
+    - `container_kvm_t`
+    - `container_logreader_t`
+    - `container_t`
+  - `level` can be anything
+  - `user` is unset
+  - `role` is unset
+
+<<[/UNRESOLVED]>>
+
+**Non-root Groups** - (restricted) This optional constraint will be omitted from the initial
+implementation.
+
+**Seccomp** - (restricted) Allow anything except `unconfined`. Same reasoning as AppArmor.
+
+<<[UNRESOLVED]>>
+
+**BLOCKING**
+
+**Capabilities** - (baseline) Adding additional capabilities beyond the default set must be
+disallowed.
+
+Kubernetes does not define the default set, but Docker does. This leaves us with 2 options:
+
+1. Hardcode the [docker default capability
+   set](https://docs.docker.com/engine/reference/run/#runtime-privilege-and-linux-capabilities), and
+   allow adding a capability in that set. This enables workloads to use an allowlist approach where
+   they drop all capabilities (`*`), and only explicitly add the ones they need. However, it means
+   that operators cannot effectively customize the default set (e.g. drop CAP_NET_RAW) through
+   runtime parameters.
+2. Forbid any use of adding capabilities. This allows runtimes to customize the default set, but
+   gives less flexibility to the pod definition.
+
+@tallclair prefers option 2
```

Comment:
NOTE: this pattern is commonly used for binding privileged ports, as in:
```
          securityContext:
            capabilities:
              drop:
                - ALL
              add:
                - NET_BIND_SERVICE
```
See https://github.com/search?l=YAML&o=desc&q=NET_BIND_SERVICE+securityContext&s=indexed&type=Code

Reply:
Note to self:
OpenShift default capability set (courtesy @mrunalp ):
```
CHOWN
DAC_OVERRIDE
FOWNER
FSETID
KILL
NET_BIND_SERVICE
SETGID
SETPCAP
SETUID
```

Docker default set:
```
+ AUDIT_WRITE
CHOWN
DAC_OVERRIDE
FOWNER
FSETID
KILL
+ MKNOD
NET_BIND_SERVICE
+ NET_RAW
+ SETFCAP
SETGID
SETPCAP
SETUID
+ SYS_CHROOT
```

I would accept the docker default set minus NET_RAW for a Kubernetes baseline set.